data/src/test/java/org/apache/iceberg/TestMergingMetrics.java

```diff
@@ -0,0 +1,164 @@
+      ID_FIELD,
+      DATA_FIELD,
+      FLOAT_FIELD,
+      DOUBLE_FIELD,
+      DECIMAL_FIELD,
+      FIXED_FIELD,
+      BINARY_FIELD,
+      FLOAT_LIST,
+      LONG_FIELD,
+      MAP_FIELD_1,
+      MAP_FIELD_2,
+      STRUCT_FIELD
+  );
+
+  protected abstract FileAppender<T> writeAndGetAppender(List<Record> records) throws Exception;
+
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
+
+  @Test
+  public void verifySingleRecordMetric() throws Exception {
+    Record record = GenericRecord.create(SCHEMA);
+    record.setField("id", 3);
+    record.setField("float", Float.NaN); // FLOAT_FIELD - 1
+    record.setField("double", Double.NaN); // DOUBLE_FIELD - 1
+    record.setField("floatlist", ImmutableList.of(3.3F, 2.8F, Float.NaN, -25.1F, Float.NaN)); // FLOAT_LIST - 2
+    record.setField("map1", ImmutableMap.of(Float.NaN, "a", 0F, "b")); // MAP_FIELD_1 - 1
+    record.setField("map2", ImmutableMap.of(
+        0, 0D, 1, Double.NaN, 2, 2D, 3, Double.NaN, 4, Double.NaN)); // MAP_FIELD_2 - 3
+
+    FileAppender<T> appender = writeAndGetAppender(ImmutableList.of(record));
+    Map<Integer, Long> nanValueCount = appender.metrics().nanValueCounts();
+
+    assertNaNCountMatch(1L, nanValueCount, FLOAT_FIELD);
+    assertNaNCountMatch(1L, nanValueCount, DOUBLE_FIELD);
+    assertNaNCountMatch(2L, nanValueCount, FLOAT_LIST);
+    assertNaNCountMatch(1L, nanValueCount, MAP_FIELD_1);
+    assertNaNCountMatch(3L, nanValueCount, MAP_FIELD_2);
+  }
+
+  private void assertNaNCountMatch(Long expected, Map<Integer, Long> nanValueCount, Types.NestedField field) {
+    Assert.assertEquals(
+        String.format("NaN count for field %s does not match expected", field.name()),
+        expected, nanValueCount.get(FIELDS_WITH_NAN_COUNT_TO_ID.get(field)));
+  }
+
+  @Test
+  public void verifyRandomlyGeneratedRecordsMetric() throws Exception {
+    List<Record> recordList = RandomGenericData.generate(SCHEMA, 50, 250L);
+
+    FileAppender<T> appender = writeAndGetAppender(recordList);
+    Map<Integer, Long> nanValueCount = appender.metrics().nanValueCounts();
+
+    FIELDS_WITH_NAN_COUNT_TO_ID.forEach((key, value) -> Assert.assertEquals(
+        String.format("NaN count for field %s does not match expected", key.name()),
+        getExpectedNaNCount(recordList, key),
+        nanValueCount.get(value)));
+
+    SCHEMA.columns().stream()
+        .filter(column -> !FIELDS_WITH_NAN_COUNT_TO_ID.containsKey(column))
+        .map(Types.NestedField::fieldId)
+        .forEach(id -> Assert.assertNull("NaN count for field %s should be null", nanValueCount.get(id)));
+  }
+
+  private Long getExpectedNaNCount(List<Record> expectedRecords, Types.NestedField field) {
+    return expectedRecords.stream()
+        .mapToLong(e -> {
+          Object value = e.getField(field.name());
+          if (value == null) {
+            return 0;
+          }
+          if (FLOAT_FIELD.equals(field)) {
+            return value.equals(Float.NaN) ? 1 : 0;
```

Comment:
Could you avoid using `equals`? Primitive NaN is not equal to itself, so it's strange to see this here. What about casting and using `Double.isNaN`?